File: wagerproof-ios-native/Wagerproof/Features/Agents/Components/AgentPerformanceCharts.swift
import SwiftUI
import Charts
import WagerproofDesign
import WagerproofModels

/// Native port of `components/agents/AgentPerformanceCharts.tsx`. Renders
/// cumulative-units line chart(s) for an agent — one overall + one per sport.
///
/// RN uses victory-native / skia for charts; we use Apple's `Charts` framework
/// which runs on UI thread Metal natively. Cumulative units is computed
/// client-side from the pick result + units/odds — same Formula B used by the
/// agent-performance recalc RPC.
struct AgentPerformanceCharts: View {
    /// Picks + parlay tickets — each settled parlay contributes exactly ONE
    /// point on the curve (its ticket-level payout), never per leg.
    let items: [AgentBetItem]
    let preferredSports: [AgentSport]
    let agentColor: Color
    /// When false the internal "Performance" heading is suppressed — used when a
    /// parent already provides a section header (the inline agent-detail layout).
    var showsTitle: Bool = true

    /// Picks-only convenience for callers without parlay data.
    init(
        allPicks: [AgentPick],
        preferredSports: [AgentSport],
        agentColor: Color,
        showsTitle: Bool = true
    ) {
        self.init(items: allPicks.map(AgentBetItem.pick),
                  preferredSports: preferredSports,
                  agentColor: agentColor,
                  showsTitle: showsTitle)
    }

    init(
        items: [AgentBetItem],
        preferredSports: [AgentSport],
        agentColor: Color,
        showsTitle: Bool = true
    ) {
        self.items = items
        self.preferredSports = preferredSports
        self.agentColor = agentColor
        self.showsTitle = showsTitle
    }

    private struct ChartPoint: Identifiable {
        // The x-index doubles as the identity — it's unique within a series and
        // stable across renders, so `Chart` diffs cheaply. (The old `UUID()`
        // per point allocated hundreds of UUIDs on every body pass — a real
        // main-thread cost for agents with long histories, and part of the
        // hitch when the chart first mounts.)
        let index: Int
        let cumulative: Double
        var id: Int { index }
    }

    private struct SportStats {
        let sport: AgentSport?
        let label: String
        let wins: Int
        let losses: Int
        let pushes: Int
        let netUnits: Double
        let points: [ChartPoint]
        /// Original x-index of the FINAL point (survives downsampling). The x-axis
        /// "Now" label keys on this, not `points.count`, so it stays correct after
        /// the render series is thinned.
        var lastIndex: Int { points.last?.index ?? 0 }
    }

    /// Everything the body needs, computed in ONE pass. Building this once (a
    /// local `let` in `body`) — instead of leaning on computed properties that
    /// re-ran the filter/sort/cumulative loop on every access — is the main fix
    /// for the "page freezes until the chart loads" hitch: the settled-items sort
    /// and per-sport curves used to be recomputed ~10× per render.
    private struct ChartModel {
        let overall: SportStats
        let bySport: [SportStats]
    }

    private func buildModel() -> ChartModel {
        let settled = items
            .filter { $0.result == .won || $0.result == .lost || $0.result == .push }
            .sorted { $0.createdAt < $1.createdAt }
        let overall = compute(label: "Overall", items: settled, sport: nil)
        // Multi-sport parlays (sportForFilter == nil) count in Overall only —
        // they don't belong to any single sport's curve.
        let bySport = preferredSports.compactMap { sport -> SportStats? in
            let scoped = settled.filter { $0.sportForFilter == sport }
            guard scoped.count >= 2 else { return nil }
            return compute(label: sport.label, items: scoped, sport: sport)
        }
        return ChartModel(overall: overall, bySport: bySport)
    }

    var body: some View {
        let model = buildModel()
        VStack(alignment: .leading, spacing: 12) {
            if model.overall.points.count < 3 {
                emptyState
            } else {
                if showsTitle {
                    Text("Performance")
                        .font(.system(size: 18, weight: .heavy))
                        .foregroundStyle(Color.appTextPrimary)
                }
                overallCard(model.overall)
                if model.bySport.count > 1 {
                    Text("By Sport")
                        .font(.system(size: 16, weight: .bold))
                        .foregroundStyle(Color.appTextPrimary)
                        .padding(.top, 4)
                    ForEach(Array(model.bySport.enumerated()), id: \.offset) { _, stats in
                        sportCard(stats: stats)
                    }
                }
            }
        }
    }

    // MARK: - Subviews

    /// Translucent glass card chrome for the chart tiles — a blurred material base
    /// under a faint dark tint, so the agent's pixelwave aura shows through
    /// instead of the old flat/opaque `appSurfaceElevated` panel (per design: the
    /// charts read as glass over the page). Shared by every chart tile below.
    private var glassCardBackground: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .fill(.ultraThinMaterial)
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .fill(Color(hex: 0x0F131C).opacity(0.5))
        }
        .overlay(
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .strokeBorder(Color.white.opacity(0.08), lineWidth: 1)
        )
    }

    private var emptyState: some View {
        VStack(spacing: 10) {
            Image(systemName: "chart.line.uptrend.xyaxis")
                .font(.system(size: 32, weight: .light))
                .foregroundStyle(Color.appTextSecondary)
            Text("Performance charts will appear after picks are graded")
                .font(.system(size: 13))
                .foregroundStyle(Color.appTextSecondary)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(28)
        .background { glassCardBackground }
    }

    private func overallCard(_ stats: SportStats) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                Text("Cumulative Units")
                    .font(.system(size: 16, weight: .bold))
                    .foregroundStyle(Color.appTextPrimary)
                Spacer()
                Text(unitsLabel(stats.netUnits))
                    .font(.system(size: 14, weight: .heavy, design: .monospaced))
                    .foregroundStyle(stats.netUnits >= 0 ? Color.appWin : Color.appLoss)
            }

            Chart(stats.points) { point in
                LineMark(
                    x: .value("Index", point.index),
                    y: .value("Units", point.cumulative)
                )
                .interpolationMethod(.monotone)
                .foregroundStyle(stats.netUnits >= 0 ? Color.appWin : Color.appLoss)
                .lineStyle(StrokeStyle(lineWidth: 2, lineCap: .round, lineJoin: .round))

                AreaMark(
                    x: .value("Index", point.index),
                    y: .value("Units", point.cumulative)
                )
                .interpolationMethod(.monotone)
                .foregroundStyle(
                    LinearGradient(
                        colors: [
                            (stats.netUnits >= 0 ? Color.appWin : Color.appLoss).opacity(0.2),
                            .clear
                        ],
                        startPoint: .top,
                        endPoint: .bottom
                    )
                )
            }
            .chartXAxis {
                AxisMarks(position: .bottom) { value in
                    AxisGridLine().foregroundStyle(Color.appBorder.opacity(0.3))
                    AxisValueLabel {
                        if let n = value.as(Int.self) {
                            Text(n == 0 ? "Start" : (n == stats.lastIndex ? "Now" : ""))
                                .font(.system(size: 10))
                        }
                    }
                }
            }
            .chartYAxis {
                AxisMarks(position: .leading) { value in
                    AxisGridLine().foregroundStyle(Color.appBorder.opacity(0.3))
                    AxisValueLabel {
                        if let v = value.as(Double.self) {
                            Text(String(format: "%+.1f", v)).font(.system(size: 10))
                        }
                    }
                }
            }
            .frame(height: 180)
        }
        .padding(14)
        .background { glassCardBackground }
    }

    private func sportCard(stats: SportStats) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                Text(stats.label)
                    .font(.system(size: 16, weight: .bold))
                    .foregroundStyle(Color.appTextPrimary)
                Spacer()
                Text("\(stats.wins)-\(stats.losses)\(stats.pushes > 0 ? "-\(stats.pushes)" : "")")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Color.appTextSecondary)
                Text(unitsLabel(stats.netUnits))
                    .font(.system(size: 13, weight: .heavy, design: .monospaced))
                    .foregroundStyle(stats.netUnits >= 0 ? Color.appWin : Color.appLoss)
            }

            if stats.points.count > 1 {
                Chart(stats.points) { point in
                    LineMark(
                        x: .value("Index", point.index),
                        y: .value("Units", point.cumulative)
                    )
                    .interpolationMethod(.monotone)
                    .foregroundStyle(stats.netUnits >= 0 ? Color.appWin : Color.appLoss)
                    .lineStyle(StrokeStyle(lineWidth: 2, lineCap: .round))
                }
                .chartXAxis(.hidden)
                .chartYAxis {
                    AxisMarks(position: .leading) { value in
                        AxisGridLine().foregroundStyle(Color.appBorder.opacity(0.25))
                        AxisValueLabel {
                            if let v = value.as(Double.self) {
                                Text(String(format: "%+.0f", v)).font(.system(size: 9))
                            }
                        }
                    }
                }
                .frame(height: 110)
            } else {
                Text("No graded picks yet")
                    .font(.system(size: 12))
                    .foregroundStyle(Color.appTextSecondary)
                    .frame(maxWidth: .infinity, minHeight: 60)
            }
        }
        .padding(14)
        .background { glassCardBackground }
    }

    // MARK: - Stats compute

    /// Upper bound on points fed to `Chart`. A cumulative-units curve is smooth,
    /// so an agent with hundreds of graded picks reads identically at ~160 points
    /// — but Swift Charts' first monotone-spline + gradient-area layout scales
    /// with point count, so thinning long histories keeps the mount snappy (the
    /// win/loss/push counts + net units are always computed from the FULL set).
    private static let maxRenderedPoints = 160

    /// Cumulative-units series via Formula B (matches `recalculate_avatar_performance`).
    /// The per-item payout math lives on `AgentBetItem.netUnitsContribution` so
    /// straight picks (odds) and parlays (settled/combined odds) share one curve.
    private func compute(label: String, items: [AgentBetItem], sport: AgentSport?) -> SportStats {
        var cumulative: Double = 0
        var points: [ChartPoint] = []
        var wins = 0, losses = 0, pushes = 0

        // Seed at 0
        points.append(ChartPoint(index: 0, cumulative: 0))

        for (idx, item) in items.enumerated() {
            switch item.result {
            case .won: wins += 1
            case .lost: losses += 1
            case .push: pushes += 1
            case .pending: continue
            }
            cumulative += item.netUnitsContribution
            points.append(ChartPoint(index: idx + 1, cumulative: cumulative))
        }

        return SportStats(
            sport: sport,
            label: label,
            wins: wins,
            losses: losses,
            pushes: pushes,
            netUnits: cumulative,
            points: Self.downsample(points, to: Self.maxRenderedPoints)
        )
    }

    /// Evenly thin a point series to at most `maxCount`, always keeping the first
    /// (seed) and last (original x-index preserved, so the "Now" axis label still
    /// lands). A no-op below the threshold.
    private static func downsample(_ pts: [ChartPoint], to maxCount: Int) -> [ChartPoint] {
        guard pts.count > maxCount, maxCount >= 2 else { return pts }
        let stride = Double(pts.count - 1) / Double(maxCount - 1)
        var out: [ChartPoint] = []
        out.reserveCapacity(maxCount + 1)
        var lastPicked = -1
        for i in 0..<maxCount {
            let idx = min(pts.count - 1, Int((Double(i) * stride).rounded()))
            if idx != lastPicked {
                out.append(pts[idx])
                lastPicked = idx
            }
        }
        // Guarantee the true final point is present (its original index backs the
        // "Now" label + the endpoint value).
        if let last = pts.last, out.last?.index != last.index {
            out.append(last)
        }
        return out
    }

    private func unitsLabel(_ n: Double) -> String {
        let sign = n >= 0 ? "+" : ""
        return String(format: "%@%.2fu", sign, n)
    }
}

/// Loading placeholder for `AgentPerformanceCharts`. Mirrors the overall card's
/// footprint — title row + 180pt chart area inside the same elevated card chrome
/// — so the swap to the real chart never jumps the layout. Detail pages render
/// this while performance picks do their first load, instead of the empty-state
/// card or a bare spinner, so the section doesn't flash empty → spinner → chart.
struct AgentPerformanceChartSkeleton: View {
    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                SkeletonBlock(width: 130, height: 15)
                Spacer()
                SkeletonBlock(width: 64, height: 13)
            }
            SkeletonBlock(height: 180, cornerRadius: 12)
        }
        .padding(14)
        .shimmering()
        // Same translucent glass chrome as the real chart tiles so the skeleton →
        // chart handoff never flips the surface opacity.
        .background {
            ZStack {
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .fill(.ultraThinMaterial)
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .fill(Color(hex: 0x0F131C).opacity(0.5))
            }
            .overlay(
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .strokeBorder(Color.white.opacity(0.08), lineWidth: 1)
            )
        }
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/components/AgentPerformanceCharts.kt
package com.wagerproof.app.features.agents.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.border
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.draw.clip
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.TextMeasurer
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.app.features.agents.agentSymbol
import com.wagerproof.core.design.components.SkeletonBlock
import com.wagerproof.core.design.components.shimmering
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.models.AgentBetItem
import com.wagerproof.core.models.AgentPick
import com.wagerproof.core.models.AgentSport
import java.util.Locale
import kotlin.math.hypot
import kotlin.math.roundToInt

/**
 * Native port of iOS `AgentPerformanceCharts`. Renders cumulative-units line
 * chart(s) for an agent — one overall + one per sport. Cumulative units is
 * computed client-side from each item's `netUnitsContribution` (Formula B, same
 * as the agent-performance recalc RPC); every settled parlay contributes exactly
 * one point (its ticket-level payout), never per leg.
 *
 * FIDELITY-WAIVER #205 (narrowed): the charts are still hand-drawn on Compose
 * Canvas rather than the iOS `Charts` framework, but the visual contract now
 * matches — monotone interpolation, axis-anchored labels, and the same 160-point
 * render cap. Only the rendering mechanism differs.
 */

// One point on a cumulative curve.
private data class ChartPoint(val index: Int, val cumulative: Double)

private data class SportStats(
    val sport: AgentSport?,
    val label: String,
    val wins: Int,
    val losses: Int,
    val pushes: Int,
    val netUnits: Double,
    val points: List<ChartPoint>,
) {
    /**
     * Original x-index of the FINAL point — survives downsampling, so the "Now"
     * axis label stays anchored to the real last pick, not to `points.size`.
     */
    val lastIndex: Int get() = points.lastOrNull()?.index ?: 0
}

/**
 * Upper bound on points fed to the Canvas. A cumulative-units curve is smooth, so
 * an 800-pick agent reads identically at ~160 points — but the Path build +
 * gradient fill scale with point count and ran on EVERY recomposition, which is
 * what made long-history agents hitch on mount (iOS caps the same way).
 */
private const val MAX_RENDERED_POINTS = 160

@Composable
fun AgentPerformanceCharts(
    items: List<AgentBetItem>,
    preferredSports: List<AgentSport>,
    // Kept for API parity with iOS; the line color is win/loss-driven, not agent-driven.
    agentColor: Color,
    showsTitle: Boolean = true,
) {
    val settledItems = remember(items) {
        items.filter {
            it.result == AgentPick.PickResultStatus.WON ||
                it.result == AgentPick.PickResultStatus.LOST ||
                it.result == AgentPick.PickResultStatus.PUSH
        }.sortedBy { it.createdAt }
    }

    val overallStats = remember(settledItems) { compute("Overall", settledItems, null) }
    val sportStats = remember(settledItems, preferredSports) {
        // Multi-sport parlays (sportForFilter == null) count in Overall only.
        preferredSports.mapNotNull { sport ->
            val scoped = settledItems.filter { it.sportForFilter == sport }
            if (scoped.size >= 2) compute(sport.label, scoped, sport) else null
        }
    }

    val textMeasurer = rememberTextMeasurer()

    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        if (overallStats.points.size < 3) {
            EmptyState()
        } else {
            if (showsTitle) {
                Text(
                    "Performance",
                    color = AppColors.appTextPrimary,
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Black,
                )
            }
            OverallCard(overallStats, textMeasurer)
            if (sportStats.size > 1) {
                Text(
                    "By Sport",
                    color = AppColors.appTextPrimary,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.padding(top = 4.dp),
                )
                sportStats.forEach { SportCard(it, textMeasurer) }
            }
        }
    }
}

// MARK: - Cards

@Composable
private fun EmptyState() {
    Column(
        modifier = Modifier.fillMaxWidth().glassCard().padding(28.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(
            agentSymbol("chart.line.uptrend.xyaxis"),
            contentDescription = null,
            tint = AppColors.appTextSecondary,
            modifier = Modifier.size(32.dp),
        )
        Text(
            "Performance charts will appear after picks are graded",
            color = AppColors.appTextSecondary,
            fontSize = 13.sp,
            textAlign = TextAlign.Center,
        )
    }
}

@Composable
private fun OverallCard(stats: SportStats, textMeasurer: TextMeasurer) {
    val accent = if (stats.netUnits >= 0) AppColors.appWin else AppColors.appLoss
    Column(
        modifier = Modifier.fillMaxWidth().glassCard().padding(14.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                "Cumulative Units",
                color = AppColors.appTextPrimary,
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
                modifier = Modifier.weight(1f),
            )
            Text(
                unitsLabel(stats.netUnits),
                color = accent,
                fontSize = 14.sp,
                fontWeight = FontWeight.Black,
                fontFamily = FontFamily.Monospace,
            )
        }
        CumulativeLineChart(
            points = stats.points,
            lastIndex = stats.lastIndex,
            color = accent,
            height = 180.dp,
            showXAxis = true,
            showYAxis = true,
            fillArea = true,
            yFormat = { String.format(Locale.US, "%+.1f", it) },
            textMeasurer = textMeasurer,
        )
    }
}

@Composable
private fun SportCard(stats: SportStats, textMeasurer: TextMeasurer) {
    val accent = if (stats.netUnits >= 0) AppColors.appWin else AppColors.appLoss
    Column(
        modifier = Modifier.fillMaxWidth().glassCard().padding(14.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                stats.label,
                color = AppColors.appTextPrimary,
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
                modifier = Modifier.weight(1f),
            )
            val record = "${stats.wins}-${stats.losses}" + if (stats.pushes > 0) "-${stats.pushes}" else ""
            Text(
                record,
                color = AppColors.appTextSecondary,
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
            )
            Spacer(Modifier.width(8.dp))
            Text(
                unitsLabel(stats.netUnits),
                color = accent,
                fontSize = 13.sp,
                fontWeight = FontWeight.Black,
                fontFamily = FontFamily.Monospace,
            )
        }
        if (stats.points.size > 1) {
            CumulativeLineChart(
                points = stats.points,
                lastIndex = stats.lastIndex,
                color = accent,
                height = 110.dp,
                showXAxis = false,
                showYAxis = true,
                fillArea = false,
                yFormat = { String.format(Locale.US, "%+.0f", it) },
                textMeasurer = textMeasurer,
            )
        } else {
            Text(
                "No graded picks yet",
                color = AppColors.appTextSecondary,
                fontSize = 12.sp,
                modifier = Modifier.fillMaxWidth().height(60.dp),
            )
        }
    }
}

// MARK: - Canvas line chart

@Composable
private fun CumulativeLineChart(
    points: List<ChartPoint>,
    lastIndex: Int,
    color: Color,
    height: Dp,
    showXAxis: Boolean,
    showYAxis: Boolean,
    fillArea: Boolean,
    yFormat: (Double) -> String,
    textMeasurer: TextMeasurer,
) {
    val labelStyle = TextStyle(color = AppColors.appTextSecondary, fontSize = 10.sp)
    Canvas(Modifier.fillMaxWidth().height(height)) {
        if (points.isEmpty()) return@Canvas

        // Insets in dp, not raw pixels: 40px is 13dp on a 3× screen, which the
        // y-axis labels overflowed straight into the plot.
        val leftPad = (if (showYAxis) 34.dp else 8.dp).toPx()
        val bottomPad = (if (showXAxis) 18.dp else 8.dp).toPx()
        val topPad = 8.dp.toPx()
        val rightPad = 8.dp.toPx()
        val labelGap = 3.dp.toPx()
        val plotW = size.width - leftPad - rightPad
        val plotH = size.height - topPad - bottomPad
        if (plotW <= 0 || plotH <= 0) return@Canvas

        val minIndex = points.first().index
        val maxIndex = points.last().index
        val idxSpan = (maxIndex - minIndex).coerceAtLeast(1).toFloat()

        var minY = points.minOf { it.cumulative }
        var maxY = points.maxOf { it.cumulative }
        if (minY == maxY) { minY -= 1.0; maxY += 1.0 }
        val ySpan = (maxY - minY)

        fun px(i: Int) = leftPad + (i - minIndex) / idxSpan * plotW
        fun py(v: Double) = topPad + (1f - ((v - minY) / ySpan).toFloat()) * plotH

        // Horizontal gridlines + Y labels (~4 ticks).
        val ticks = 4
        for (t in 0..ticks) {
            val v = minY + ySpan * t / ticks
            val y = py(v)
            drawLine(
                AppColors.appBorder.copy(alpha = 0.3f),
                Offset(leftPad, y),
                Offset(size.width - rightPad, y),
                strokeWidth = 1f,
            )
            if (showYAxis) {
                val layout = textMeasurer.measure(yFormat(v), labelStyle)
                drawText(layout, topLeft = Offset(2.dp.toPx(), y - layout.size.height / 2f))
            }
        }

        // Vertical gridlines (iOS AxisMarks(position: .bottom) draws one per tick).
        if (showXAxis) {
            val xTicks = 4
            for (t in 0..xTicks) {
                val x = leftPad + plotW * t / xTicks
                drawLine(
                    AppColors.appBorder.copy(alpha = 0.3f),
                    Offset(x, topPad),
                    Offset(x, topPad + plotH),
                    strokeWidth = 1f,
                )
            }
        }

        val offsets = points.map { Offset(px(it.index), py(it.cumulative)) }

        // Area fill under the curve.
        if (fillArea && offsets.size > 1) {
            val baseline = topPad + plotH
            val area = monotonePath(offsets).apply {
                lineTo(offsets.last().x, baseline)
                lineTo(offsets.first().x, baseline)
                close()
            }
            drawPath(
                area,
                brush = Brush.verticalGradient(
                    colors = listOf(color.copy(alpha = 0.2f), Color.Transparent),
                    startY = topPad,
                    endY = baseline,
                ),
            )
        }

        // The line.
        if (offsets.size > 1) {
            drawPath(
                monotonePath(offsets),
                color = color,
                style = Stroke(width = 2f * density),
            )
        }

        // X labels, anchored to the first/last RENDERED point (their original
        // indices survive downsampling) rather than to fixed pixel offsets.
        if (showXAxis) {
            val labelY = size.height - bottomPad + labelGap
            val startLayout = textMeasurer.measure("Start", labelStyle)
            drawText(
                startLayout,
                topLeft = Offset(
                    (px(minIndex) - startLayout.size.width / 2f).coerceIn(0f, size.width - startLayout.size.width),
                    labelY,
                ),
            )
            val nowLayout = textMeasurer.measure("Now", labelStyle)
            drawText(
                nowLayout,
                topLeft = Offset(
                    (px(lastIndex) - nowLayout.size.width / 2f).coerceIn(0f, size.width - nowLayout.size.width),
                    labelY,
                ),
            )
        }
    }
}

/**
 * Monotone cubic Hermite (Fritsch–Carlson) through the offsets — the same shape
 * iOS gets from `.interpolationMethod(.monotone)`. The old quadratic-through-
 * midpoints curve could bulge past a local max/min, drawing a units peak the
 * agent never actually reached.
 */
private fun monotonePath(pts: List<Offset>): Path {
    val path = Path()
    if (pts.isEmpty()) return path
    path.moveTo(pts.first().x, pts.first().y)
    val n = pts.size
    if (n < 2) return path
    if (n == 2) {
        path.lineTo(pts[1].x, pts[1].y)
        return path
    }

    val dx = FloatArray(n - 1)
    val slope = FloatArray(n - 1)
    for (i in 0 until n - 1) {
        dx[i] = pts[i + 1].x - pts[i].x
        slope[i] = if (dx[i] == 0f) 0f else (pts[i + 1].y - pts[i].y) / dx[i]
    }

    val m = FloatArray(n)
    m[0] = slope[0]
    m[n - 1] = slope[n - 2]
    for (i in 1 until n - 1) {
        // A sign flip is a local extremum: flatten the tangent so the spline
        // touches it instead of overshooting through it.
        m[i] = if (slope[i - 1] * slope[i] <= 0f) 0f else (slope[i - 1] + slope[i]) / 2f
    }
    for (i in 0 until n - 1) {
        if (slope[i] == 0f) {
            m[i] = 0f
            m[i + 1] = 0f
            continue
        }
        val a = m[i] / slope[i]
        val b = m[i + 1] / slope[i]
        val h = hypot(a, b)
        if (h > 3f) {
            val t = 3f / h
            m[i] = t * a * slope[i]
            m[i + 1] = t * b * slope[i]
        }
    }
    for (i in 0 until n - 1) {
        val third = dx[i] / 3f
        path.cubicTo(
            pts[i].x + third, pts[i].y + m[i] * third,
            pts[i + 1].x - third, pts[i + 1].y - m[i + 1] * third,
            pts[i + 1].x, pts[i + 1].y,
        )
    }
    return path
}

// MARK: - Stats compute

/** Cumulative-units series via Formula B (matches recalculate_avatar_performance). */
private fun compute(label: String, items: List<AgentBetItem>, sport: AgentSport?): SportStats {
    var cumulative = 0.0
    val points = mutableListOf(ChartPoint(index = 0, cumulative = 0.0))
    var wins = 0
    var losses = 0
    var pushes = 0

    items.forEachIndexed { idx, item ->
        when (item.result) {
            AgentPick.PickResultStatus.WON -> wins++
            AgentPick.PickResultStatus.LOST -> losses++
            AgentPick.PickResultStatus.PUSH -> pushes++
            AgentPick.PickResultStatus.PENDING -> return@forEachIndexed
        }
        cumulative += item.netUnitsContribution
        points.add(ChartPoint(index = idx + 1, cumulative = cumulative))
    }

    return SportStats(
        sport = sport,
        label = label,
        wins = wins,
        losses = losses,
        pushes = pushes,
        netUnits = cumulative,
        // Counts + net units always come from the FULL set; only the drawn
        // series is thinned.
        points = downsample(points, MAX_RENDERED_POINTS),
    )
}

/**
 * Evenly thin a series to at most [maxCount] points, always keeping the first
 * (seed) and last (its original index backs the "Now" axis label). No-op below
 * the threshold.
 */
private fun downsample(pts: List<ChartPoint>, maxCount: Int): List<ChartPoint> {
    if (pts.size <= maxCount || maxCount < 2) return pts
    val stride = (pts.size - 1).toDouble() / (maxCount - 1)
    val out = ArrayList<ChartPoint>(maxCount + 1)
    var lastPicked = -1
    for (i in 0 until maxCount) {
        val idx = minOf(pts.size - 1, (i * stride).roundToInt())
        if (idx != lastPicked) {
            out.add(pts[idx])
            lastPicked = idx
        }
    }
    val last = pts.last()
    if (out.lastOrNull()?.index != last.index) out.add(last)
    return out
}

private fun unitsLabel(n: Double): String {
    val sign = if (n >= 0) "+" else ""
    return String.format(Locale.US, "%s%.2fu", sign, n)
}

/**
 * Translucent glass chrome for the chart tiles. The iOS `.ultraThinMaterial`
 * blur isn't reproduced here (FIDELITY-WAIVER #301) — a dark tint + faint white
 * hairline stands in so the agent aura still reads through.
 */
private fun Modifier.glassCard(corner: Dp = 16.dp): Modifier = this
    .clip(RoundedCornerShape(corner))
    .background(Color(0xFF0F131C).copy(alpha = 0.5f))
    .border(1.dp, Color.White.copy(alpha = 0.08f), RoundedCornerShape(corner))

/**
 * Loading placeholder mirroring the overall card footprint (title row + 180dp
 * chart area inside the same glass chrome) so the swap to the real chart never
 * jumps the layout.
 */
@Composable
fun AgentPerformanceChartSkeleton() {
    Column(
        modifier = Modifier.fillMaxWidth().glassCard().padding(14.dp).shimmering(),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            SkeletonBlock(width = 130.dp, height = 15.dp)
            Spacer(Modifier.weight(1f))
            SkeletonBlock(width = 64.dp, height = 13.dp)
        }
        SkeletonBlock(height = 180.dp, cornerRadius = 12.dp, modifier = Modifier.fillMaxWidth())
    }
}
